--- timidity/timidity/Makefile.am.omv~	2014-07-29 19:47:56.523067695 +0200
+++ timidity/timidity/Makefile.am	2014-07-29 19:48:03.509781834 +0200
@@ -18,6 +18,8 @@
 
 include $(top_srcdir)/common.makefile.in
 
+LIBS += -laudio
+
 INCLUDES = \
 	-I$(top_srcdir) \
 	-I$(top_srcdir)/libarc \
--- timidity/timidity/mfi.c.omv~	2014-07-29 19:47:44.069649515 +0200
+++ timidity/timidity/mfi.c	2014-07-29 19:47:50.743028363 +0200
@@ -344,7 +344,7 @@ typedef struct LastNoteInfo {
 #define SEND_LASTNOTEINFO(lni, ch)				if (LASTNOTEINFO_HAS_DATA((lni)[ch])) SendLastNoteInfo(lni, ch);
 #define SEND_AND_CLEAR_LASTNOTEINFO(lni, ch)	if (LASTNOTEINFO_HAS_DATA((lni)[ch])) { SendLastNoteInfo(lni, ch); (lni)[ch].on = NO_LAST_NOTE_INFO; }
 
-inline void StoreLastNoteInfo(LastNoteInfo *info, int channel, int time, int duration, int note, int velocity)
+static void StoreLastNoteInfo(LastNoteInfo *info, int channel, int time, int duration, int note, int velocity)
 {
 	info[channel].on = time;
 	info[channel].off = time + duration;
@@ -352,7 +352,7 @@ inline void StoreLastNoteInfo(LastNoteIn
 	info[channel].velocity = velocity;
 }
 
-inline void SendLastNoteInfo(const LastNoteInfo *info, int channel)
+static void SendLastNoteInfo(const LastNoteInfo *info, int channel)
 {
 	NOTE_BUF_EV_DEBUGSTR(channel, info[channel].on, note_name[info[channel].note % 12], info[channel].note / 12, info[channel].velocity, info[channel].off);
 	MIDIEVENT(info[channel].on, ME_NOTEON, channel, info[channel].note, info[channel].velocity);
