--- TiMidity++-2.13.2/configure.in.emacs	2005-02-16 01:14:04.726491000 +0800
+++ TiMidity++-2.13.2/configure.in	2005-02-16 02:50:33.549456840 +0800
@@ -1602,7 +1602,8 @@
   [ ELFILES="$ELFILES timidity.el"
     INTERFACE_SRCS="$INTERFACE_SRCS emacs_c.c"
   ],
-  [ ELFILES="$ELFILES timidity.el" ])
+  [ ELFILES="$ELFILES timidity.el"
+    AM_CONDITIONAL(ENABLE_EMACS, true)])
 
 dnl VT100
 AM_CONDITIONAL(ENABLE_VT100, false)
--- TiMidity++-2.13.2/interface/timidity.el.emacs	2002-01-15 18:53:20.000000000 +0800
+++ TiMidity++-2.13.2/interface/timidity.el	2005-02-16 03:22:06.513682536 +0800
@@ -35,7 +35,7 @@
 
 ;; Configuration parameters.
 ; Absolute path of timidity.
-(defvar timidity-prog-path "/usr/local/bin/timidity")
+(defvar timidity-prog-path "/usr/bin/timidity")
 
 ; String list for timidity program options.
 (defvar timidity-default-options nil)
